internal/client/client.go

```diff
@@ -148,38 +148,84 @@ func (c *CSAPI) SyncUntilInvitedTo(t *testing.T, roomID string) {
 			event.Get("state_key").Str == c.UserID
 	}
-	c.SyncUntil(t, "", "", "rooms.invite."+GjsonEscape(roomID)+".invite_state.events", check)
+	c.SyncUntilArray(t, "", "", "rooms.invite."+GjsonEscape(roomID)+".invite_state.events", check)
 }
 
-// SyncUntil blocks and continually calls /sync until
+// SyncUntilArray blocks and continually calls /sync until
 // - the response contains a particular `key`, and
 // - its corresponding value is an array
 // - some element in that array makes the `check` function return true.
-// If the `check` function fails the test, the failing event will be automatically logged.
+// If the `check` function fails the test, the failing value will be automatically logged.
 // Will time out after CSAPI.SyncUntilTimeout.
-func (c *CSAPI) SyncUntil(t *testing.T, since, filter, key string, check func(gjson.Result) bool) {
-	t.Helper()
-	start := time.Now()
-	checkCounter := 0
+func (c *CSAPI) SyncUntilArray(t *testing.T, since, filter, key string, check func(gjson.Result) bool) {
+	var wasFailed *bool
+	var timedOut *bool
+	var checkCounter *int
+	var lastElement *gjson.Result
+
 	// Print failing events in a defer() so we handle t.Fatalf in the same way as t.Errorf
-	var wasFailed = t.Failed()
-	var lastEvent *gjson.Result
-	timedOut := false
 	defer func() {
-		if !wasFailed && t.Failed() {
+		// When failing on a fatal error
+		if !*wasFailed && t.Failed() {
 			raw := ""
-			if lastEvent != nil {
-				raw = lastEvent.Raw
+			if lastElement != nil {
+				raw = lastElement.Raw
 			}
-			if !timedOut {
-				t.Logf("SyncUntil: failing event %s", raw)
+			if !*timedOut {
+				t.Logf("SyncUntilArray: failing element %s", raw)
 			}
 		}
 	}()
+
+	arrayCheck := func(result gjson.Result) bool {
+		if result.IsArray() {
+			elements := result.Array()
+
+			for i, el := range elements {
+				lastElement = &el
+				if check(el) {
+					return true
+				}
+				*wasFailed = t.Failed()
+				// Increment counter artificially for every sub-check call
+				// Makes syncUntilInternal report "correctly" how many checks were made
+				if i != 0 {
+					*checkCounter++
+				}
+			}
+		}
+		return false
+	}
+
+	c.syncUntilInternal(t, since, filter, key, wasFailed, timedOut, checkCounter, arrayCheck)
+}
+
+// SyncUntil blocks and continually calls /sync until
+// - the response contains a particular `key`, and
+// - the corresponding value makes the `check` function return true.
+// Will time out after CSAPI.SyncUntilTimeout.
+func (c *CSAPI) SyncUntil(t *testing.T, since, filter, key string, check func(gjson.Result) bool) {
+	var wasFailed *bool
+	var timedOut *bool
+	var checkCounter *int
+
+	c.syncUntilInternal(t, since, filter, key, wasFailed, timedOut, checkCounter, check)
+}
+
+// Internal function helping both SyncUntil and SyncUntilArray with some logging coordination upon failure
+func (c *CSAPI) syncUntilInternal(t *testing.T, since, filter, key string, wasFailed, timedOut *bool, checkCounter *int, check func(gjson.Result) bool) {
```

Comment:
This function signature is a bit bonkers, but essentially, most of these are used by `SyncUntilArray` to "feed back" the values in its `defer`'d function (as `t.Fatalf` just stops the goroutine like a panic and runs all defers up the stack, and i wanted to separate `SyncUntil` and `SyncUntilArray` functionality.